### PR TITLE
fix(dot/state): change map of tries implementation to have working garbage collection

### PR DESCRIPTION
**Pull request by qdm12**
## Changes

- [x] Remove `syncing` field (always false) from `StorageState`
- [x] Add new type `tries` implementing a mutex locked map of root hash to trie pointers

## Tests

```
go test ./dot/state
```

## Issues

- #1973 

## Primary Reviewer

- @noot

### CODE REVIEW

File: dot/state/storage.go [hunk 1]
```diff
@@ -71,36 +70,22 @@ func NewStorageState(db chaindb.Database, blockState *BlockState,
...
 	root := ts.MustRoot()
 
-	if s.syncing {
-		// keep only the trie at the head of the chain when syncing
-		// TODO: probably remove this when memory usage improves (#1494)
-		s.tries.Range(func(k, _ interface{}) bool {
-			s.tries.Delete(k)
-			return true
-		})
-	}
```
Comment (qdm12) [inline dot/state/storage.go:101]: this was never run and looks rather strange, so I removed it

File: dot/state/roottottrie.go [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+	defer r.Unlock()
+
+	var calculatedRoot common.Hash
+	t, has := r.rootToTrie[root]
+	if has {
+		calculatedRootHash := t.MustHash()
+		if !calculatedRootHash.Equal(root) {
+			panic(fmt.Sprintf("trie does not have expected root, expected %s but got %s",
+				root, calculatedRootHash))
+		}
```
Comment (qdm12) [inline dot/state/roottottrie.go:69]: Can I remove those panics (here and below)? Or is there a risk of a root hash mismatch here? :thinking: 

That might accelerate processing since we don't need to encode and hash the entire trie (`t.MustHash()`)

File: dot/node.go [hunk 3]
```diff
@@ -177,13 +176,6 @@ func LoadGlobalNodeName(basepath string) (nodename string, err error) {
 
 // NewNode creates a new dot node from a dot node configuration
 func NewNode(cfg *Config, ks *keystore.GlobalKeystore) (*Node, error) {
-	// set garbage collection percent to 10%
-	// can be overwritten by setting the GOGC env variable, which defaults to 100
-	prev := debug.SetGCPercent(10)
-	if prev != 100 {
-		debug.SetGCPercent(prev)
-	}
-
```
Comment (noot) [inline dot/node.go:186]: why is this removed? does it not affect memory usage anymore?

File: dot/state/roottottrie.go (hunk 2 above)
Comment (noot) [reply dot/state/roottottrie.go:69]: yeah, you can remove them, in the case where the trie gets corrupted for whatever reason and the roots don't match, the block processing higher up will fail with a state root mismatch

**Review by noot — APPROVED** [review]
nice cleanup!

File: dot/state/roottottrie.go [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+)
+
+type rootToTrieMap struct {
+	rootToTrie map[common.Hash]*trie.Trie
+	sync.RWMutex
+}
```
Comment (noot) [inline dot/state/roottottrie.go:18]: I'm curious as to the advantage of this over a *sync.Map

File: dot/node.go (hunk 3 above)
Comment (qdm12) [reply dot/node.go:186]: I would like to remove it to have a clear picture. I don't really think it helps much anyway, the GC is pretty optimized. And also our memory problem is not GC related. Is there a reason we added it in the first place?

Comment (noot) [reply dot/node.go:186]: the memory usage was way worse without it, but we can remove it and just set it if needed

File: dot/state/roottottrie.go (hunk 4 above)
Comment (qdm12) [reply dot/state/roottottrie.go:18]: Just type safety / cleaner code. That's how I found out about the empty pruned keys so I thought I would push it 😄

I also don't think it matters much since we just read and access it not in the million per second AFAIK.
If we identify it as a bottleneck later, then we can change it to a sync.Map, which is really for corner performance cases I would say. Let's not pre-pre-optimize 😉

File: dot/node.go (hunk 3 above)
Comment (qdm12) [reply dot/node.go:186]: We can always set it with `GOGC` if you need to run the node for as long as you want.
Although I guess setting to 10 will bottleneck the CPU so the program will run slower, hence the longer time perhaps? 🤔 
From my point of view, I prefer to have it run as fast as possible and crash to see if my changes fixed it or not 😄 So I guess I can also use `GOGC`, up to you I can also revert it!

File: dot/state/roottottrie.go (hunk 2 above)
Comment (qdm12) [reply dot/state/roottottrie.go]: So I removed the check when loading from memory since it's rather not possible to have it corrupt I think. I left the one when loading the trie from disk.

By the way, I ran gossamer with a print to see how often it would load the trie from disk in this function and it looks like it never does (for 30 min), is this expected?

Comment (noot) [reply dot/state/roottottrie.go]: yeah, it shouldn't really ever load from disk except on start-up, so that sounds about right

Comment (noot) [reply dot/state/roottottrie.go]: it's rather slow to load from disk so we don't want that when syncing

File: dot/node.go (hunk 3 above)
Comment (noot) [reply dot/node.go:186]: no, it's fine to delete it, we can just re-add it later if we want

File: dot/state/roottottrie.go (hunk 2 above)
Comment (qdm12) [reply dot/state/roottottrie.go]: Ok that makes sense. Maybe some simplification could be done, since it likely loads from disk from another code path so this code path in this function is never really run. Anyway, in another PR later maybe some day in the future.

File: dot/state/roottottrie.go [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+	r.rootToTrie[root] = trie
+}
+
+func (r *rootToTrieMap) delete(root common.Hash) {
+	r.Lock()
+	defer r.Unlock()
+	delete(r.rootToTrie, root)
+}
+
+func (r *rootToTrieMap) getFromMemOrDB(root common.Hash, db chaindb.Database) (
```
Comment (timwu20) [inline dot/state/roottottrie.go:57]: can we move this code into the caller of this function `StorageState.TrieState()`. Doesn't make a lot of sense for the in-memory `rootToTrieMap` has to worry about offline storage `chaindb.Database`.

File: dot/state/storage.go [hunk 6]
```diff
@@ -30,7 +30,7 @@ func errTrieDoesNotExist(hash common.Hash) error {
 // StorageState is the struct that holds the trie, db and lock
 type StorageState struct {
 	blockState *BlockState
-	tries      *sync.Map // map[common.Hash]*trie.Trie // map of root -> trie
+	rootToTrie *rootToTrieMap
```
Comment (timwu20) [inline dot/state/storage.go:33]:
```suggestion
	tries *rootToTrieMap
```
nit: can we keep the same syntax.  I kinda like `s.tries`.

Comment (qdm12) [reply dot/state/storage.go]: done in d7b3f3ffe256a3aa9fa602c5ddd35a414fe177a8

File: dot/state/roottottrie.go (hunk 5 above)
Comment (qdm12) [reply dot/state/roottottrie.go:57]: >> Doesn't make a lot of sense...

Actually it does, I think it should be an atomic operation to get from memory + get from db + set to memory.

Otherwise we could end up with cases where

- goroutine A gets root hash XYZ from memory, it's not found
- goroutine B sets trie at root hash XYZ in memory
- goroutine A gets stored trie for root hash XYZ from disk
- goroutine A sets trie for root hash XYZ in memory

Effectively rendering the write of B erased. Probably won't happen, but better safe 😉

File: dot/state/roottottrie.go (hunk 4 above)
Comment (qdm12) [reply dot/state/roottottrie.go:18]: It doesn't suck at garbage collection

File: dot/state/roottottrie.go (hunk 5 above)
Comment (timwu20) [reply dot/state/roottottrie.go:57]: Then shouldn't `chaindb` be stored in `rootToTrieMap` instead of being passed on every call to `getFromMemOrDB`.  Also after looking at the calling code in `StorageState` it looks it already checks the map if it has the key, and will load from db in `StorageState.loadTrie`, and it only needs this atomic read in `StorageState.TrieState`.  If it were me, I'd simplify this and just add `Lock()`, `Unlock()` methods to the `rootToTrieMap`.

File: dot/state/roottottrie.go [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+import (
+	"fmt"
+	"sync"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+)
+
+type rootToTrieMap struct {
```
Comment (timwu20) [inline dot/state/roottottrie.go:15]:
```suggestion
type trieMap struct {
```

File: dot/state/roottottrie.go [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+	r.Lock()
+	defer r.Unlock()
+
+	_, exists := r.rootToTrie[root]
+	if exists {
+		return
+	}
+
+	r.rootToTrie[root] = trie
+}
```
Comment (EclesioMeloJunior) [inline dot/state/roottottrie.go:49]: I would suggest to use the already written function because they use diferent locks for reading and writing
```suggestion
	_, exists := r.get(root)
	if exists { return }
	r.set(root, trie)
}
```

File: dot/state/roottottrie.go [hunk 9]
```diff
@@ -0,0 +1,88 @@
...
+func (r *rootToTrieMap) delete(root common.Hash) {
+	r.Lock()
+	defer r.Unlock()
+	delete(r.rootToTrie, root)
+}
+
+func (r *rootToTrieMap) getFromMemOrDB(root common.Hash, db chaindb.Database) (
+	t *trie.Trie, err error) {
+	r.Lock()
+	defer r.Unlock()
```
Comment (EclesioMeloJunior) [inline dot/state/roottottrie.go:60]:
```suggestion
```

File: dot/state/roottottrie.go [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+
+	var calculatedRoot common.Hash
+	t, has := r.rootToTrie[root]
+	if has {
+		return t, nil
+	}
```
Comment (EclesioMeloJunior) [inline dot/state/roottottrie.go:66]: use the `get` to use the inner read locks
```suggestion
	t, has := r.get(root)
	if has {
		return t, nil
	}
```

File: dot/state/roottottrie.go [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+		return nil, fmt.Errorf("cannot load root %s from database: %w", root, err)
+	}
+
+	calculatedRootHash := t.MustHash()
+	if !calculatedRootHash.Equal(root) {
+		panic(fmt.Sprintf("trie does not have expected root, expected %s but got %s",
+			root, calculatedRootHash))
+	}
+
+	r.rootToTrie[calculatedRoot] = t
```
Comment (EclesioMeloJunior) [inline dot/state/roottottrie.go:80]:
```suggestion
	r.set(calculatedRoot, t)
```

**Review by EclesioMeloJunior — CHANGES_REQUESTED** [review]

File: dot/state/roottottrie.go (hunk 8 above)
Comment (qdm12) [reply dot/state/roottottrie.go:49]: It has to be atomic, this won't work (get is atomic, set is atomic but not both)

File: dot/state/roottottrie.go (hunk 10 above)
Comment (qdm12) [reply dot/state/roottottrie.go:66]: the whole operation has to be thread safe, this is not possible 😉

File: dot/state/roottottrie.go (hunk 11 above)
Comment (qdm12) [reply dot/state/roottottrie.go:80]: the whole operation has to be thread safe, this is not possible 😉

File: dot/state/roottottrie.go (hunk 9 above)
Comment (qdm12) [reply dot/state/roottottrie.go:60]: the whole operation has to be thread safe, this is not possible 😉

File: dot/state/tries.go [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+func (t *tries) getValue(trieRoot common.Hash, key []byte) (
+	value []byte, err error) {
+	// Try to get from memory
+	t.mapMutex.RLock()
+	tr, has := t.rootToTrie[trieRoot]
+	t.mapMutex.RUnlock()
+	if has {
+		value = tr.Get(key)
+		// TODO: should we return an error if the value is not found?
+		return value, nil
```
Comment (qdm12) [inline dot/state/tries.go:44]: Question here! ❓ 

When a key is not found in the database, we return an error. However, from memory we don't, should we change that? Or return a `nil` value from the database if not found?

File: dot/state/roottottrie.go (hunk 7 above)
Comment (qdm12) [reply dot/state/roottottrie.go]: Changed it to `tries` since it now has the in memory map and a reference to the persistent database.

File: dot/state/roottottrie.go (hunk 5 above)
Comment (qdm12) [reply dot/state/roottottrie.go]: Changed pretty much a lot of code in there; now more operations are atomic and the newer `tries` structure handles the operations of memory + database.

File: dot/state/roottottrie.go (hunk 2 above)
Comment (qdm12) [reply dot/state/roottottrie.go]: I changed it to be used in more places, now the read from disk code path is indeed called. 👍 
That way I also spotted a child trie test bug with the root hash panic check when reading from disk, although I'm still working on debugging it.

File: dot/state/tries.go (hunk 12 above)
Comment (EclesioMeloJunior) [reply dot/state/tries.go:44]: @qdm12 I was looking at the [GetFromDB](https://github.com/ChainSafe/gossamer/blob/6c64521f57f24556489cb2183bf19b1e2f02ab79/lib/trie/database.go#L280) and [getFromDB](https://github.com/ChainSafe/gossamer/blob/6c64521f57f24556489cb2183bf19b1e2f02ab79/lib/trie/database.go#L306), the functions whose retrieves the data from database, and the returned errors means problems while getting data (eg. database closed or database doesn't exists) or decoding the data (eg. incomplete or wrong encoded hashes), however, if we can retrieve and decode the hashes and even though the key is not present both functions will return nil without error, so I think we don't need to change the behavior. Please, correct me if I'm wrong

File: dot/state/roottottrie.go (hunk 5 above)
Comment (qdm12) [reply dot/state/roottottrie.go]: Moved changes regarding atomic operations in #2250

File: dot/state/tries.go (hunk 12 above)
Comment (qdm12) [reply dot/state/tries.go]: Awesome thanks! Not sure why I thought it was returning an error when not found...

I also moved changes regarding atomic operations in #2250 so I'll remove that `TODO`

File: lib/trie/database.go [hunk 13]
```diff
@@ -279,7 +279,7 @@ func (t *Trie) ClearPrefixFromDB(db chaindb.Database, prefix []byte) error {
 // It then reads the value from the database.
 func GetFromDB(db chaindb.Database, rootHash common.Hash, key []byte) (
 	value []byte, err error) {
-	if rootHash == EmptyHash {
+	if len(key) == 0 && rootHash == EmptyHash {
```
[the remark this answers is not in the document]
Comment (qdm12) [reply lib/trie/database.go]: Totally. Changed back in #2250 , not sure why I did that in the first place 🤔

File: dot/state/roottottrie.go (hunk 2 above)
Comment (qdm12) [reply dot/state/roottottrie.go]: Root hash check moved to PR #2251 (with child trie test failing for now)

**Review by EclesioMeloJunior — APPROVED** [review]
